Milestone 2: Frienships V1

Mindo-Joseph opened the pull request:
#### What does this PR do?
- Implements friendships association.

#### Description of Task to be completed?
- Create a model with associations and all requested features for friendships

#### How should this be manually tested?
- In the root directory of the project run ``` rspec spec/features``` to run the feature tests and ```rspec spec/models``` to run model specs

Mindo-Joseph committed:
Create friendship model, check if friendship relation exists between users

edieatha committed:
Add other helper methods for friendships

edieatha committed:
Fixed error

Mindo-Joseph committed:
Modify gemfile and create test for validating user is created

Mindo-Joseph committed:
Create association tests for user

edieatha committed:
Add tests for posts model

edieatha committed:
Add tests for comments model

Mindo-Joseph committed:
Create tests for like

Mindo-Joseph committed:
Test associations of friendship model

edieatha committed:
Create auth_spec.rb added tests

Mindo-Joseph committed:
Test login page

Mindo-Joseph committed:
Test the signup feature

Mindo-Joseph committed:
Fix lint errors

Mindo-Joseph committed:
Fix more lint errors

Mindo-Joseph committed:
Update rubocop file

Mindo-Joseph committed:
Remove unrecognized cop

edieatha committed:
Update user_spec.rb

edieatha committed:
Fix linter error

maelfosso commented on app/models/user.rb:
```diff
@@ -9,4 +9,30 @@ class User < ApplicationRecord
+  has_many :friendships
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: :friend_id
+
+  def friends
+    friends_array = friendships.map { |friendship| friendship.friend if friendship.confirmed }
+    friends_array << inverse_friendships.map { |friendship| friendship.friend if friendship.confirmed }
+    friends_array.compact
+  end
```
Can be done in a better way
```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

maelfosso commented on app/models/user.rb:
```diff
@@ -9,4 +9,30 @@ class User < ApplicationRecord
+    friends_array.compact
+  end
+
+  def pending_friends
+    friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
+  end
+
+  def friend_requests
+    inverse_friendships.find { |friendship| friendship.user unless friendship.confirmed }.compact
+  end
```
You can update it by following the same method as the ones suggested for **friends** up there.

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

maelfosso commented on app/models/user.rb:
```diff
@@ -9,4 +9,30 @@ class User < ApplicationRecord
+    inverse_friendships.find { |friendship| friendship.user unless friendship.confirmed }.compact
+  end
+
+  def confirm_friend(user)
+    friendship = inverse_friendships.find { |friend| friend.user == user }
+    friendship.confirmed = true
+    friendship.save
+  end
```
If you implement that why didn't you use it?

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello,

Thank you for the effort put into this project.
Please, a more bit a effort to close this project and move forward.

1. Fix the linter's rrors

2. Posts timeline must only include posts written by logged-in user and their friends
![Screenshot at 2020-07-22 06:32:55](https://user-images.githubusercontent.com/8848104/88138249-337cfd00-cbe5-11ea-9447-7d56de4f2048.png)
But from this screenshot, **Bob Bar** and **Alice Bar** are not yet friends.

3.
I'm connected as **Bob Bar**
![Screenshot at 2020-07-22 06:34:44](https://user-images.githubusercontent.com/8848104/88138387-7c34b600-cbe5-11ea-9ea4-2b766520452a.png)
On this screenshot, It's the **Alice Bar** profile but it seems missing a button.
It's required that you put a button to invite user to friendship on users list page and on single user page.
But there is none on the users' list page too as you can see here.
![Screenshot at 2020-07-22 06:37:30](https://user-images.githubusercontent.com/8848104/88138582-e0f01080-cbe5-11ea-9bbf-fea22a210ff3.png)

4.
Don't forget that the main features in this milestone is **friendship request**

5. 
Break your **custom.css** in many parts to take advantage of Ruby on Rails framework.

6.
`rspec` generated this error
```rb
An error occurred while loading ./spec/models/user_spec.rb.
Failure/Error: return load_without_bootsnap(resolved, wrap)

SyntaxError:
  /home/maelfosso/Documents/Microverse/TSE/RoR/stay-in-touch/spec/models/user_spec.rb:28: syntax error, unexpected end-of-input, expecting `end'
# /home/maelfosso/.rvm/gems/ruby-2.7.1/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
# /home/maelfosso/.rvm/gems/ruby-2.7.1/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'


Finished in 0.00004 seconds (files took 4.93 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```

7. 
Kindly read the comments below

Make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

Mindo-Joseph committed:
Modify user model to use through tables for association. Add tests for the new associations

Mindo-Joseph committed:
Add link to become friends, fix the friend checker method

edieatha committed:
Created friendship controller

Add resources for friendship controller
Add Create method to controller

edieatha committed:
Add friendship_status partial view

edieatha committed:
Fixed friendship model and features

Added custom style

edieatha committed:
Timeline will now display only friends posts

Non friend posts wont be viewed!

edieatha commented:
##  TSE Recommendations applied
✔️ Add a fully working friend: request, reject, accept, unfriend feature
✔️ Add Timeline feature where only posts from friends are viewed by current user.

Mindo-Joseph committed:
Fix rubocop cop name

Mindo-Joseph committed:
Fix lint errors

Mindo-Joseph committed:
Remove extra empty lines

Mindo-Joseph committed:
Update tests for user model

Mindo-Joseph committed:
Remove extra empty line

kobiyoyo reviewed (changes requested):
## Hi Team    ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, but there are a few more things required. 
### General
- [ ] I get this error whenever I click on `see profile`
![1](https://user-images.githubusercontent.com/23278161/88335007-8a401f00-cd2a-11ea-9c1d-4735c993c667.png)
- [ ] I should be able to delete or accept friend request 
- [ ] Fix linter errors 
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

edieatha commented:
* [x]  The error is not present on my environment! Were the migrations ran? It seems to work flawlesly...
   * [x]  You should be able to delete or accept friend request
   * [x]  Fixed linter errors

![Screenshot_2020-07-24 RorSocialScaffold](https://user-images.githubusercontent.com/59659283/88338716-6fc57000-cd41-11ea-974a-6a1f9b0652cc.png)

edieatha committed:
Additional styling and refactor

majovanilla commented on app/models/user.rb:
```diff
@@ -9,4 +9,22 @@ class User < ApplicationRecord
+  has_many :pending_friendships, -> { where confirmed: nil }, class_name: 'Friendship', foreign_key: 'friend_id'
+
+  def friends
+    sent_request = Friendship.where(user_id: id, confirmed: true).pluck(:friend_id)
+    received_request = Friendship.where(friend_id: id, confirmed: true).pluck(:user_id)
+    ids = sent_request + received_request
```
Like the first reviewer suggested, there are better ways to implement this method.
Using operations (+) is not a good practice. That's why you have the associations. Please implement them.

majovanilla commented on app/models/user.rb:
```diff
@@ -9,4 +9,22 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+
+  has_many :friendships
```
You are missing the other side of the friendship (inverse_friendship) association.

majovanilla commented on app/views/users/show.html.erb:
```diff
@@ -1,9 +1,38 @@
+	  </li><br>
+	  
+	<div class="card col-sm-9">
     <h3>Recent posts:</h3>
     <ul class="posts">
       <%= render @posts %>
     </ul>
```
You shouldn't show the user's posts if they are not friends.

majovanilla commented on app/views/users/index.html.erb:
```diff
@@ -2,10 +2,11 @@
+        <%= user.name %></br>
+          <%= link_to 'See Profile',  user_path(user), class: 'btn btn-outline-success btn-sm' %>
+        <% if signed_in? && !Friendship.reacted?(current_user.id, user.id) && current_user != user %>
+          <td> <%= link_to 'Add Friend', friendships_create_path(ids: {id1: current_user.id, id2: user.id}), class: 'btn btn-success btn-sm' %></td>
+		<% end %>
```
When you add a friend, you hide the button but you don't show the `pending_invitation` button. Please include it.

majovanilla commented on app/views/users/show.html.erb:
```diff
@@ -1,9 +1,38 @@
+	    <% @user.pending_friendships.each do |friendship| %>
+	      <li>
+	        <%= User.find(friendship.user_id).name %>
+	        <%= link_to 'Accept', friendships_update_path(friendship_id: friendship.id),  class: 'btn btn-success btn-sm'  %>
+	        <%= link_to 'Reject', friendships_destroy_path(friendship_id: friendship.id),  class: 'btn btn-danger btn-sm' %>
```
**Optional**. These links are visible only in the user's profile. It's not very user-friendly because you have to click All users and then the profile. It would be better to give immediate access to the current user profile with a link in the navbar.

You are not showing any messages of success or failure when accepting a friend or deleting one.

majovanilla reviewed (changes requested):
# Changes required :mag: 


Hey team, 👋

## Good job on this project! 👍 You are almost there!!

There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Please check the comments I left on the code and implement all of them. 

### Optional Changes:

If you want to make your project even better 🔥💥, I have a few suggestions you can implement.  

These are completely [Optional]: 🦔

- Please check the comments marked as optional in the code below.
- It would be nice to see the User's name on the navbar.

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

edieatha committed:
Add current user profile to navbar

Additional styling

Mindo-Joseph committed:
1. Use associations to determine friends instead of a function
2. Include inverse_friendships association
3. Only show posts from friends
4. Upon sending an invite request, a pending_invitation button is displayed

Mindo-Joseph committed:
Fix lint errors

Mindo-Joseph committed:
Remove pending_friendships association

Mindo-Joseph committed:
Add tests for confirmed friends and inverse_friendships

mosesogwo reviewed (changes requested):
**Hello!:clap: I see great effort and great progress:+1:, but there are still a few issues:**
:point_right:  User show page is not accessible

![image](https://user-images.githubusercontent.com/46611870/88421149-efa21780-cddf-11ea-8c67-0c2351bab940.png)

:point_right: Accepting a friend a request is not working. It has "invitation pending" on both sides of the association. The sender should have invitation pending and the receiver should have Accept and Reject options.

:point_right: No tests have been added for the friendship feature.

_**Please reach out to me on Slack (Moses Ogwo) when you are available so we can go through your associations together. The user associations through friendships are not complete**_

**Cheers:stars:**

### Status: Changes Requested.

edieatha committed:
Restore pending  friendships fix user profile page

edieatha committed:
Fixed invitattion pending on both sides of the association, accept friend request now works

edieatha committed:
Restored old model of associations

edieatha committed:
Fixed user model associations

edieatha committed:
Changed confirmed default value,

edieatha committed:
Can't see user posts if not friends

edieatha committed:
Fixed requestor shouldn't see receiver's requests from same

Mindo-Joseph committed:
Add tests for friendship

Mindo-Joseph committed:
Resolve stickler errors:

Mindo-Joseph committed:
Resolve stickler errors:

uche-inyama reviewed (approved):
## Approved

Hi team :wave: 

This project is approved.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```
